src/reconstruction.jl

```diff
@@ -0,0 +1,214 @@
+using Statistics
+using LinearAlgebra
+export AbstractSpatialEmbedding
+export SpatioTemporalEmbedding, STE
+export outdim
+export AbstractBoundaryCondition, PeriodicBoundary, ConstantBoundary
+
+import DynamicalSystemsBase: reconstruct
+
+#####################################################################################
+#                 Spatio Temporal Delay Embedding Reconstruction                    #
+#####################################################################################
+"""
+    AbstractSpatialEmbedding <: AbstractEmbedding
+Super-type of spatiotemporal embedding methods.
+Use `subtypes(AbstractSpatialEmbedding)` for available methods.
+"""
+abstract type AbstractSpatialEmbedding{T,Φ,BC,X} <: AbstractEmbedding end
+const ASE = AbstractSpatialEmbedding
+"""
+    AbstractBoundaryCondition
+Super-type of boundary conditions for [`SpatioTemporalEmbedding`](@ref).
+Use `subtypes(AbstractBoundaryCondition)` for available methods.
+"""
+abstract type AbstractBoundaryCondition end
+
+"""
+	ConstantBoundary(C)
+Constant boundary condition type. Enforces constant boundary conditions
+when passed to [`SpatioTemporalEmbedding`](@ref)
+by filling missing out-of-bounds values in the reconstruction with
+parameter `C`.
+"""
+struct ConstantBoundary{T} <: AbstractBoundaryCondition
+    C::T
+end
+
+"""
+	PeriodicBoundary
+Periodic boundary condition struct. Enforces periodic boundary conditions
+when passed to [`SpatioTemporalEmbedding`](@ref) in the reconstruction.
+"""
+struct PeriodicBoundary    <: AbstractBoundaryCondition end
+
+
+"""
+	Region{Φ}
+Internal struct for efficiently keeping track of region far from boundaries of field.
+Used to speed up reconstruction process.
+"""
+struct Region{Φ}
+	mini::NTuple{Φ,Int}
+	maxi::NTuple{Φ,Int}
+end
+
+Base.length(r::Region{Φ}) where Φ = prod(r.maxi .- r.mini .+1)
+function Base.in(idx, r::Region{Φ}) where Φ
+	for φ=1:Φ
+		r.mini[φ] <= idx[φ] <= r.maxi[φ] || return false
+ 	end
+ 	return true
+end
+Base.CartesianIndices(r::Region{Φ}) where Φ =
+	 CartesianIndices{Φ,NTuple{Φ,UnitRange{Int}}}(([r.mini[φ]:r.maxi[φ] for φ=1:Φ]...,))
+
+
+function inner_region(βs::Vector{CartesianIndex{Φ}}, fsize) where Φ
+	mini = Int[]
+	maxi = Int[]
+	for φ = 1:Φ
+		js = map(β -> β[φ], βs) # jth entries
+		mi,ma = extrema(js)
+		push!(mini, 1 - min(mi, 0))
+		push!(maxi,fsize[φ] - max(ma, 0))
+	end
+	return Region{Φ}((mini...,), (maxi...,))
+end
+
+function project_inside(α::CartesianIndex{Φ}, r::Region{Φ}) where Φ
```

Comment:
Is this function allocating? I think it can be made non-allocating if it is allocating.

Reply:
```
julia> α = CartesianIndex(1,2,3,4,5)
CartesianIndex(1, 2, 3, 4, 5)

julia> r = TimeseriesPrediction.Region((1,1,1,1,1),(3,3,3,3,3))
TimeseriesPrediction.Region{5}((1, 1, 1, 1, 1), (3, 3, 3, 3, 3))

julia> @btime TimeseriesPrediction.project_inside($α, $r)
  22.651 ns (0 allocations: 0 bytes)
CartesianIndex(1, 2, 3, 1, 2)
```